Description: Patch from upstream to fix crashes in the resolver.
Origin: upstream, http://hg.debian.org/hg/aptitude/head/rev/b1438b17e637
Bug: http://bugs.debian.org/578344
Forwarded: not-needed
Author: Daniel Burrows <dburrows@debian.org>
Applied-Upstream: http://hg.debian.org/hg/aptitude/head/rev/b1438b17e637
Index: aptitude-0.6.2/src/generic/problemresolver/promotion_set.h
===================================================================
--- aptitude-0.6.2.orig/src/generic/problemresolver/promotion_set.h	2010-04-19 06:54:44.180978398 -0700
+++ aptitude-0.6.2/src/generic/problemresolver/promotion_set.h	2010-04-19 06:55:07.828278340 -0700
@@ -151,8 +151,16 @@
 
     // Note that this will compute a somewhat inefficient validity
     // condition when applied across several expressions.
-    cwidget::util::ref_ptr<expression<bool> > new_valid =
-      and_e::create(p1_valid, p2_valid);
+    cwidget::util::ref_ptr<expression<bool> > new_valid;
+    if(p1_valid.valid())
+      {
+        if(p2_valid.valid())
+          new_valid = and_e::create(p1_valid, p2_valid);
+        else
+          new_valid = p1_valid;
+      }
+    else
+      new_valid = p2_valid;
 
     return generic_promotion(new_choices, new_cost, new_valid);
   }
